File: enigmajs/src/rotor.js
import {rotors} from './config';

const CHAR_COUNT = 26;
const CODE_OFFSET = 65;

export class Rotor {
    constructor(settings) {
        this.position = settings.position;
        this.type = settings.type;
        this.rmap = rotors[settings.type].map;
    }

    shift(letter) {
        let poffset = this.position.charCodeAt(0) - CODE_OFFSET;
        let loffset = letter.charCodeAt(0) - CODE_OFFSET;

        let code = poffset + loffset;

        if (code >= CHAR_COUNT) {
            code = code - CHAR_COUNT;
        }

        return this.rmap[code];
    }

    unshift(letter) {
        let poffset = this.position.charCodeAt(0) - CODE_OFFSET;
        let offset = this.rmap.indexOf(letter);

        let code = offset - poffset;

        if (code < 0) {
            code = CHAR_COUNT + code;
        }

        return String.fromCharCode(code + CODE_OFFSET);
    }

    move() {
        if (this.position === 'Z') {
            this.position = 'A';
            return;
        }

        let pos = this.position.charCodeAt(0) - CODE_OFFSET;
        this.position = String.fromCharCode(CODE_OFFSET + ++pos);
    }
}

File: enigmarust/enigmawasm/src/rotor.rs
use std::char;
use config::Config;

pub struct Rotor {
    pub position: char,
    rotor_map: Vec<char>,
    pub step: char
}

impl Rotor {
    pub fn create(rtype: String, position: String, config: &Config) -> Rotor {
        let rconfig = config.rotors.get(&rtype.clone()).unwrap();

        Rotor {
            position: position.chars().next().unwrap(),
            rotor_map: rconfig.map.clone().chars().collect(),
            step: rconfig.step.clone()
        }
    }

    pub fn shift(&self, letter: char) -> char {
        let poffset: u8 = (self.position as u8) - 65;
        let loffset: u8 = (letter as u8) - 65;

        let mut code = poffset + loffset;

        if code >= 26 {
            code = code - 26;
        }

        self.rotor_map[code as usize]
    }

    pub fn unshift(&self, letter: char) -> char {
        let poffset: i32 = (self.position as i32) - 65;
        let offset: i32 = self.rotor_map.iter().position(|&c| c == letter).unwrap() as i32;

        let mut code = poffset - offset;

        if code < 0 {
            code = 26 + code;
        }

        char::from_u32((code as u32) + 65).unwrap()
    }

    pub fn move_rotor(&mut self) {
        if self.position == 'Z' {
            self.position = 'A';
            return;
        }

        self.position = char::from_u32((self.position as u32) + 1).unwrap();
    }
}